### PR TITLE
Support Ember-CLI-Mocha users.

### LINKED ISSUE #6: QUnit-flavor tests generated for non qunit users
At the very least, if they're not using qunit, we should warn and not generate a test.

```js
import Ember from 'ember';
import { moduleFor, test }
from 'ember-qunit';

moduleFor('validator:better-confirmation', 'Unit | Validator | better-confirmation', {
  needs: ['validator:messages']
});

test('it works', function(assert) {
  var validator = this.subject();
  assert.ok(validator);
});
```

### PR DESCRIPTION
**Pull request by blimmer**
Fixes #6

### CODE REVIEW

**Comment by blimmer** [comment]
Lodash blueprinting is ugly - here's what a default test for validator `foo-bar` looks like:

```js
/* jshint expr:true */
import { expect } from 'chai';
import Ember from 'ember';
import {
  describeModule,
  it
} from 'ember-mocha';

describeModule(
  'validator:foo-bar',
  'FooBarValidator',
  {
    // Specify the other units that are required for this test.
    needs: ['validator:messages']
  },
  function() {
    it('works', function() {
      const validator = this.subject();
      expect(validator).to.be.ok;
    });
  }
);

```

**Comment by stefanpenner** [comment]
it should be possible for us to just choose different files based on the current test runner. That may be nicer then creating branching in the same file.

Thoughts?

**Comment by blimmer** [comment]
Would love to see how to do that - do you have an example?

**Comment by blimmer** [comment]
Reminder for me to sync with potentially changing this to an integration test based on [this convo](https://github.com/offirgolan/ember-cp-validations/pull/10#discussion_r39039430) .

**Comment by blimmer** [comment]
@stefanpenner are you thinking something like [this](https://github.com/ember-cli/ember-cli/blob/master/blueprints/resource/index.js)? Doesn't seem obvious how to override install for choosing a different file per test-type.

**Comment by blimmer** [comment]
@rwjblue do you know any secret sauce on how I could have two distinct flavors of a testing file and ember-cli chooses the right one based on which testing framework you use? I've seen [a lot](https://github.com/ember-cli/ember-cli/blob/master/blueprints/component-test/files/tests/__testType__/__path__/__test__.js) of [examples](https://github.com/minutebase/ember-can/blob/master/blueprints/ability-test/index.js#L25) of people doing it this way, but maybe there's some feature I'm missing?

**Comment by rwjblue** [comment]
Key things: 

* Keep the test blueprint separate from the main one. In this case, it means adding a `validator-test` blueprint which is automatically invoked by `ember generate` command for you if it exists ([here](https://github.com/ember-cli/ember-cli/blob/1bdff397b725a7820eaeb09a120b949db0294c56/lib/tasks/generate-from-blueprint.js#L14)). This allows you to avoid duplicating the validator code while having different types of test blueprints.
* Make the `validator-test` blueprint check if `ember-cli-qunit` or `ember-cli-mocha` is installed ([like this](https://github.com/minutebase/ember-can/blob/ef64960b3c925ea66dc9ec1ba5d07325b37afeae/blueprints/ability-test/index.js#L7-L13))
* Override [`files` function](https://github.com/ember-cli/ember-cli/blob/1bdff397b725a7820eaeb09a120b949db0294c56/lib/models/blueprint.js#L291-L302) in `validator-test` blueprint to detect the test type, and use an appropriate files directory.

This might look something like:

```javascript
// blueprints/validator-test/index.js
var path = require('path');
var existsSync = require('exists-sync');
var walkSync = require('walk-sync');

module.exports = {
  description: 'Generates a validator unit test'

  _getFilesPath: function() {
    var type;
    if ('ember-cli-mocha' in this.project.addonPackages) {
      type = 'mocha';
    } else if ('ember-cli-qunit' in this.project.addonPackages) {
      type = 'qunit';
    }
    return path.join(this.path, type + '-files');
  },

  files: function() {
    if (this._files) { return this._files; }

    var filesPath = this._getFilesPath();

    if (existsSync(filesPath)) {
      this._files = walkSync(filesPath);
    } else {
      this._files = [];
    }

    return this._files;
  },

  srcPath: function(file) {
    return path.resolve(this._getFilesPath(), file);
  }
};
```

At this point, you can have a folder structure like:

```
blueprints/validator/index.js
blueprints/validator/files/**/*

blueprints/validator-test/index.js

blueprints/validator-test/qunit-files/**/*

blueprints/validator-test/mocha-files/**/*
```

---

I much prefer this over the being forced to template the files in JS (via strings) as done [here](https://github.com/minutebase/ember-can/blob/ef64960b3c925ea66dc9ec1ba5d07325b37afeae/blueprints/ability-test/index.js#L25-L61).

**Comment by rwjblue** [comment]
As an aside, I believe that the above customizations could either be moved upstream into ember-cli or just made into a small helper/utility library that folks could use.

**Comment by rwjblue** [comment]
I submitted https://github.com/ember-cli/ember-cli/pull/4837 to make this simpler in future ember-cli versions.

**Comment by blimmer** [comment]
Thanks rwjblue!

**Comment by blimmer** [comment]
This is updated and ready for review.

File: blueprints/validator/files/app/validators/__name__.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
-import Ember from 'ember';
 import BaseValidator from 'ember-cp-validations/validators/base';
 
 export default BaseValidator.extend({
-  validate(value, options, model, attribute) {
+  validate(/*value, options, model, attribute*/) {
     return true;
```
Comment (blimmer) [inline blueprints/validator/files/app/validators/__name__.js]: Default .jshint on ember-cli projects warns on unused attributes.

File: blueprints/validator-test/qunit-files/tests/unit/validators/__name__-test.js [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import Ember from 'ember';
```
Comment (blimmer) [inline blueprints/validator-test/qunit-files/tests/unit/validators/__name__-test.js]: Default .jshint on ember-cli projects warns on unused attributes.

File: blueprints/validator-test/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    } else if ('ember-cli-qunit' in this.project.addonPackages) {
+      type = 'qunit';
+    }
+    return path.join(this.path, type + '-files');
+  },
+
+  // The files and srcPath method can be removed once
+  // https://github.com/ember-cli/ember-cli/pull/4837 is live.
+  // Also, remove the npm dependencies on walk-sync and exists-sync when fixing.
+  files: function() {
```
Comment (rwjblue) [inline blueprints/validator-test/index.js]: Can we future proof this, and only override if on ember-cli before 1.13.9? We are already including ember-cli-version-checker for detecting Ember version and adding ember-get-helper when < Ember 2.0.

File: blueprints/validator-test/index.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+var path = require('path');
+var existsSync = require('exists-sync');
+var walkSync = require('walk-sync');
+
+module.exports = {
+  description: 'Generates a validator unit test',
+
+  _getFilesPath: function() {
```
Comment (rwjblue) [inline blueprints/validator-test/index.js]: Let's name this filesPath to be in line with future versions once that PR lands.

File: blueprints/validator-test/index.js (hunk 3 above)
Comment (rwjblue) [inline blueprints/validator-test/index.js]: In that case, we would detect in init if we need to use overeidden files and srcPath functions, and if we do we would set them in init. Possibly just via something like this:

    this.files = this._legacyEmberCLIFilesOveride;

Comment (stefanpenner) [inline blueprints/validator-test/index.js]: should we sniff that the functionality exists? detecting some method?

Comment (rwjblue) [inline blueprints/validator-test/index.js]: Sure, we could check in init if this.filesPath exists and if it does, replace it with our custom one above. If it doesn't then we replace this.files and this.srcPath like I mentioned above.

File: blueprints/validator-test/index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    var type;
+    if ('ember-cli-mocha' in this.project.addonPackages) {
+      type = 'mocha';
+    } else if ('ember-cli-qunit' in this.project.addonPackages) {
+      type = 'qunit';
+    }
```
Comment (offirgolan) [inline blueprints/validator-test/index.js]: If the app isnt using mocha or qunit, we should either default to qunit or exit out with a message.

**Comment by blimmer** [comment]
Sorry for the delay on this - this is fixed up according to comments.

**Comment by rwjblue** [comment]
:+1: - LGTM

**Pull request merged**